### ISSUE TITLE
1.0 Plans

### ISSUE THREAD

**Issue by yuloh** [thephpleague/json-guard#98]
I wrote a whole road map for 1.0 in the github issue tab before my browser crashed :(

Here is a rough outline of what I'm planning on doing.  I did a little bit of this work on master before I decided I should just plan for 1.0.

### Simpler Constraints

Every constraint implements a single interface.  See #88.  The interface also requires an instance now (non static method), which makes writing constraints with outside dependencies easier.

These changes were partly to make writing a custom rule set for swagger easier.  That will hopefully be released as a separate package shortly after 1.0 is released.

### Simpler Internationalization

Every context array will always have the keys `value`, `parameter` to make writing language files simpler.  Right now the parameter is named specific to the rule, i.e `enum` => `choices`.

The constraint keyword constants are moving from the constraint class to a single file.  The idea is to save you from having to check every class to see what keywords need language lines.  It also means we don't have to have an inconsistent constant name when the constraint validates multiple keywords (like `max/exclusiveMax`).

I'm not sure what to do about language files.  I'm leaning towards making a separate repo of lang files that can be used with symfony/translation.

### Better Errors

- Errors will have a pointer to where in the schema the constraint is, and a copy of the schema.  #86 

- Errors for container instance constraints (`oneOf`, `allOf` etc) will include the errors of the sub schema that failed .  #82 

### Separate Reference Library

The dereferencing code was starting to take over the library.  It's being moved to a separate library and will be pulled in as a dependency.  This was originally discussed in #95.  There are a couple of changes:

- Support for caching schemas using PSR-16 (SimpleCache) implementations.  The reference was also changed to optimize the size of the serialized schema, and it no longer uses a closure to resolve.  An in memory cache is about 5x faster.  Improvements with something like redis varies depending on the complexity of the schema and if you were already connected.  Sometimes it's slower.  This should help a lot for schemas that use http references.

- The dereferencer has an interface to allow decorating (i.e. the cache decorator).

- Swappable scope resolvers.  JSON Schema requires you to resolve a reference against any `id` keywords in the schema.  You can now inject your own implementation to resolve the scope differently or not specify an argument to disable this behaviour.

- True proxies.  References are now iterable so you can use them (almost) like real objects.  As long as your code isn't using get_object_vars your code doesn't even have to know it's using references.

- Loader Manager.  All of the code for registering and getting loaders was moved to a separate class.

### Dropping Support For 5.5 and HHVM

HHVM builds take 5x longer to finish than PHP builds.  I'm honestly never going to have the time to install hhvm and figure out why.  HHVM's lack of [type hints on variadics](https://github.com/facebook/hhvm/issues/6954) means I can't use dev dependencies that would be useful.

5.5 has been EOL for 7 months now.  The library uses a lot of namespaced functions and so being able to import functions would be a big plus for readability :+1  I would also like to be able to require-dev phpbench which is 5.6+

**Comment by yuloh** [comment]
I'm merging 1.0 features into [the one-o branch](https://github.com/thephpleague/json-guard/tree/one-o) until it's released.

**Comment by moderndeveloperllc** [comment]
What kind of BC breaks do you foresee with 1.0 beyond the library split?

**Comment by yuloh** [comment]
I'm not 100% sure at the moment.  I started an upgrade guide section in the documentation, so I will keep that updated as I merge stuff into the 1.0 branch.

If you are only using the validator for validating Draft4 JSON Schema with the default ruleset and constraints, the BC breaks should be minimal.  I imagine most users will just have to update the use statement for the dereferencer and everything will work.

If you are using custom constraints, you will need to update them to use the simpler interface.  The constructor for the error returned by constraints is probably going to change a bit, so you would need to update that too.

There are also some changes to how registering custom loaders works, since that was moved to a separate class.  That will only affect users who aren't using the default file and http(s) loaders.

**Comment by handrews** [comment]
@yuloh do you have any plans to support draft-06?  It should be published literally any day now (the "draft-05" spec which didn't have a meta-schema expires on the 16th of this month).

**Comment by yuloh** [comment]
Hey @handrews 👋 

I already added support for the `$id` keyword.  I just saw that @epoberezkin finished merging all the tests so I will start working on support for the rest of draft 6 support soon.

**Comment by handrews** [comment]
@yuloh Awesome, thanks!  I'm just trying to get a feel for who might be giving draft-06 a try.

**Comment by yuloh** [comment]
I just tagged 1.0.0-alpha, so I'm closing this issue.

**Comment by marcioAlmada** [comment]
@yuloh so what would the be state on internalization here?

I had internationalization for the `0.*` json-league in place but with the changes in the data structure it's not compatible anymore. Before any rework on my part, is there any facility for internationalization planned?

**Comment by yuloh** [comment]
Hi @marcioAlmada,

The errors were changed to always use the [same keys in the context array](https://github.com/thephpleague/json-guard/blob/master/src/ValidationError.php#L178) so internationalization should be much simpler now.  Every constraint is also in it's own file and has a constant for the [keyword](https://github.com/thephpleague/json-guard/blob/master/src/Constraint/DraftFour/ExclusiveMaximum.php#L12).  Most error messages just use `value` and `parameter` but if those don't fully describe the reason for the failure they use `cause` [like this](https://github.com/thephpleague/json-guard/blob/master/src/Constraint/DraftFour/Required.php#L30).

Official support for internationalization would be great.  With the current state of the library it should be possible to add without any changes to the library itself.  I was imagining a translator being a separate component but I suppose the validator could translate errors as it collects them.  Either way it should be pretty straightforward to add.

I created issue #129 to track internationalization.  Any input on how you think it should work or a PR is definitely welcome 😁
